File: mobile/feature-headlines/src/main/java/com/rogoz208/feature_headlines/presentation/articles_list/recycler/ArticlesDiffCallback.java
package com.rogoz208.feature_headlines.presentation.articles_list.recycler;

import androidx.recyclerview.widget.DiffUtil;


import com.rogoz208.mobile_common.model.ArticleUI;

import java.util.List;
import java.util.Objects;

public class ArticlesDiffCallback extends DiffUtil.Callback {

    private final List<ArticleUI> oldList;
    private final List<ArticleUI> newList;

    public ArticlesDiffCallback(List<ArticleUI> oldList, List<ArticleUI> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return Objects.equals(oldList.get(oldItemPosition).getId(), newList.get(newItemPosition).getId());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return Objects.equals(oldList.get(oldItemPosition).getDescription(), newList.get(newItemPosition).getDescription())
                && Objects.equals(oldList.get(oldItemPosition).getSourceName(), newList.get(newItemPosition).getSourceName())
                && Objects.equals(oldList.get(oldItemPosition).getUrlToImage(), newList.get(newItemPosition).getUrlToImage())
                && Objects.equals(oldList.get(oldItemPosition).getSourceLogoUrl(), newList.get(newItemPosition).getSourceLogoUrl());
    }
}

File: mobile/feature-sources/src/main/java/com/rogoz208/feature_sources/articles_screen/recycler/ArticlesDiffCallback.kt
package com.rogoz208.feature_sources.articles_screen.recycler

import androidx.recyclerview.widget.DiffUtil
import com.rogoz208.mobile_common.model.ArticleUI


class ArticlesDiffCallback(
    private val oldList: List<ArticleUI>,
    private val newList: List<ArticleUI>
) : DiffUtil.Callback() {
    override fun getOldListSize() = oldList.size

    override fun getNewListSize() = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].id == newList[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].description == newList[newItemPosition].description
                && oldList[oldItemPosition].sourceName == newList[newItemPosition].sourceName
                && oldList[oldItemPosition].urlToImage == newList[newItemPosition].urlToImage
                && oldList[oldItemPosition].sourceLogoUrl == newList[newItemPosition].sourceLogoUrl
    }
}
